vxblastsms/ussd.py

```diff
@@ -89,83 +56,98 @@ def handle_raw_inbound_message(self, message_id, request):
             )
             return
 
-        from_addr = values['msisdn']
-        provider = self.normalise_provider(values['provider'])
-        ussd_session_id = optional_values['ussdSessionId']
-
-        if optional_values['to_addr'] is not None:
+        if values['type'] == '2':  # resume session
             session_event = TransportUserMessage.SESSION_RESUME
-            to_addr = optional_values['to_addr']
-            content = optional_values['request']
+        elif values['type'] == '1':  # new session
+            session_event = TransportUserMessage.SESSION_NEW
         else:
+            # TODO: handle other types if necessary
+            # Default to new session for 3 & 4 for now
             session_event = TransportUserMessage.SESSION_NEW
-            to_addr = optional_values['request']
+
+        if optional_values['msg'] is not None:
+            content = optional_values['msg']
+        else:
             content = None
 
         log.info(
-            'BlastSMSUssdTransport receiving inbound message from %s to %s.' % (
-                from_addr, to_addr))
+            'BlastSMSUssdTransport receiving inbound message from %s to '
+            '%s.' % (values['msisdn'], values['shortcode']))
 
         yield self.publish_message(
             message_id=message_id,
             content=content,
-            to_addr=to_addr,
-            from_addr=from_addr,
+            to_addr=values['shortcode'],
+            from_addr=values['msisdn'],
+            # provider=None,
             session_event=session_event,
             transport_type=self.transport_type,
+            transport_name=self.transport_name,
             transport_metadata={
-                'aat_ussd': {
-                    'provider': provider,
-                    'ussd_session_id': ussd_session_id,
-                }
+                'sessionid': values['sessionid'],
+                'appid': optional_values['appid'],
+
             },
-            provider=provider,
         )
 
-    def generate_body(self, reply, callback, session_event):
-        request = Element('request')
-        headertext = SubElement(request, 'headertext')
-        headertext.text = reply
+    def generate_body(self, msisdn, sessionid, appid, in_reply_to,
+                      reply_content, session_event):
+
+        e_ussdresp = Element('ussdresp')
+
+        se_msisdn = SubElement(e_ussdresp, 'msisdn')
+        se_msisdn.text = msisdn
 
-        # If this is not a session close event, then send options
+        se_sessionid = SubElement(e_ussdresp, 'sessionid')
+        se_sessionid.text = sessionid
+
+        se_appid = SubElement(e_ussdresp, 'appid')
+        if appid is not None:
+            se_appid.text = appid
+        else:
+            # provide application session id - use reference to inbound
+            # message that started the session
+            se_appid.text = in_reply_to
```

Comment:
I'd rather we set this to an optional configurable value at the transport level. Or we leave it as not set.